### DISCUSSION TITLE
The truth about MaxiBlocks performance optimization

### DISCUSSION THREAD

**Discussion by ckp267** [maxi-blocks/maxi-blocks#5228] [Announcements]
> I have read in different Facebook groups that MaxiBlocks is not really performance optimized, e.g. it adds more divs than necessary to the HTML code.

Do the lighthouse speed test here: https://maxiblocks.com/ and be sure to compare with pages that have the same amout of blocks, images, interactions, hovers, buttons AND 6 responsive break points including 4K to 360px screens. Few builders perform this well.

<img width="1919" alt="MaxiBlocks com_Lighthouse_97" src="https://github.com/maxi-blocks/maxi-blocks/assets/7437580/57eade0e-8edd-44f7-838f-408f9530a5a1">

**Comment by elzadj** [comment]
The divs in html don't really affect performance, it's js and css + php and server response time (mostly js and css).

https://stackoverflow.com/questions/6106219/does-the-use-of-multiple-divs-affect-page-performance

We have 200-300 divs for the biggest pages. That's not a significant amount at all. 

This page has 786 divs https://www.elegantthemes.com/gallery/divi/

This page has 432 divs https://time.com

448 https://www.ulah.es/en/

More info:

1. Parsing Time: Every additional element will take some time for the browser to parse. However, the time required to parse a single <div> or even several of them is extremely minimal. Modern browsers are highly optimized for this, and the difference in parsing time between a page with, say, 100 <div> elements and one with 150 <div> elements would be negligible.
3. DOM Size: As mentioned previously, a very large DOM can slow down a page. But you would need thousands of additional nodes to see a noticeable difference in performance. If you're only adding a reasonable number of static <div> elements, it's unlikely to have any significant impact on performance.
4. Memory Consumption: Each DOM element consumes some memory. But, again, the memory consumed by a single <div> or even dozens of them is very small. You'd need an extraordinarily large number of <div> elements to run into memory-related issues on modern devices.

In practice, for a typical web page, adding or removing a few <div> elements won't have a noticeable impact on performance. Even adding several hundred would likely be imperceptible in terms of page load speed on modern hardware and browsers.

In conclusion, while it's a good practice to avoid unnecessary markup for the sake of clarity and maintainability, the performance impact of static <div> elements on their own is negligible. It becomes a concern only when their count goes into the high thousands, which is an uncommon scenario for most websites.

**Comment by ckp267** [comment]
### WP Hive test is not good?

The WP hive test is not an accurate assessment because it combines metrics for both the **Editing Experience (EE)** backend-end the **Visitor Experience (VE)** front end into one combined metric. https://wphive.com/plugins/maxi-blocks/

**This is not representative of how is used in practice.** It's not accurate to the final output on the front end VE pages IE. Lighthouse metrics. 

### Let me explain

- Gutenberg does all the heavy lifting in the EE plus with MaxiBlocks we add a ton of components to our blocks (90+). These components like the "dynamic content component" is exactly the reason we are able to create such unique patterns and templates. 
- Please take a look https://maxiblocks.com/demo/ and make a fair apples-for-apples comparison with any competing block and template library. We spent 3.8 years developing this. 
- And for the amount of functionality we add in the back-end EE it's as snappy as you can hope for with the feature set - all free btw. Speed is everything. Still, there's always room to improve and we will.
- Next point is that we also have PHP saving attributes to the server - to reduce the load in the browser AND save block styles to the database for Full Site Editing. 
- This enables "Theme builder functionality" headers, footers, blogs, archives and so on. So the EE uses more memory to give you all the features, which then gets output to the VE as only the necessary html, css and script files (if needed).

**WP hive should ideally update their tests to be closer to actual use cases** - or make new tests for block based plugins and themes. The metric you should look at is the **/frontpage/** - we actually make it faster! See screenshot.

<img width="852" alt="MaxiBlocks_Wp_hive_V1 5 1" src="https://github.com/maxi-blocks/maxi-blocks/assets/7437580/ef07100b-506b-43a4-97a0-ba6f3c81b72f">

### Lastly, as they say, Walk the talk.
Here's a video I just recorded on instawp.io with MaxiBlocks installed, with WordPress 2023 theme. 

- It shows how fast it loads. I also show all the components in the right sidebar - that's just the text block - which is one of most feature rich. Then I do a quick Lighthouse speed test for that pattern. I think the results speak for themselves. 

**Anyone can launch the demo here too** https://app.instawp.io/launch?t=maxiblocks-demo-v151&d=v2

https://github.com/maxi-blocks/maxi-blocks/assets/7437580/f5beeeb2-098f-4ec4-9036-872d9009cc1c

Let me know if you have more comments or questions.
